### PR TITLE
[attention] fix test for packed padfree masking

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

As per title, I added the test a few days ago as slow, so I can fix them later for all models

### CODE REVIEW

File: tests/test_modeling_common.py [hunk 1]
```diff
@@ -4177,55 +4193,48 @@ def flash_attention_padding_matches_padding_free_with_position_ids(
...
-                # We need to do simple forward without cache in roder to trigger packed SDPA/FLEX/EAGER path
+                # We need to do simple forward without cache in order to trigger packed SDPA/flex/eager attention path
                 res_padded = model(**inputs_dict, use_cache=False)
                 res_padfree = model(**padfree_inputs_dict, use_cache=False)
 
-                logits_padded = res_padded.logits[inputs_dict["attention_mask"].bool()]
+                logits_padded = res_padded.logits[dummy_attention_mask.bool()]
                 logits_padfree = res_padfree.logits[0]
 
-                torch.testing.assert_close(logits_padded.argmax(-1), logits_padfree.argmax(-1), rtol=0, atol=0)
```
Comment (zucchini-nlp) [inline tests/test_modeling_common.py:4218]: flaky otherwise, even tiny changes in logits result in different sampled tokens

File: src/transformers/models/chameleon/modeling_chameleon.py [hunk 2]
```diff
@@ -43,9 +43,7 @@
 
 
 if is_torch_flex_attn_available():
-    from torch.nn.attention.flex_attention import BlockMask
-
-    from ...integrations.flex_attention import make_flex_block_causal_mask
+    pass
```
Comment (vasqu) [inline src/transformers/models/chameleon/modeling_chameleon.py]: Why not remove entirely?

File: tests/models/emu3/test_modeling_emu3.py [hunk 3]
```diff
@@ -234,9 +234,9 @@ def prepare_config_and_inputs(self):
         config = self.get_config()
 
         input_ids = ids_tensor([self.batch_size, self.seq_length], config.text_config.vocab_size)
-        attention_mask = input_ids.ne(1).to(torch_device)
         input_ids[input_ids == self.image_token_id] = self.pad_token_id
         input_ids[:, : self.image_seq_length] = self.image_token_id
+        attention_mask = input_ids.ne(1).to(torch_device)
```
Comment (vasqu) [inline tests/models/emu3/test_modeling_emu3.py]: Does the order matter here? I'd guess pad and img token are not 1?

File: tests/models/chameleon/test_modeling_chameleon.py [hunk 4]
```diff
@@ -270,7 +270,7 @@ def prepare_config_and_inputs(self):
         input_ids = ids_tensor([self.batch_size, self.seq_length], self.vocab_size)
         input_ids[input_ids == self.image_token_id] = self.pad_token_id
         input_ids[:, : self.image_seq_length] = self.image_token_id
-        attention_mask = torch.tril(torch.ones_like(input_ids).to(torch_device))
+        attention_mask = input_ids.ne(self.pad_token_id).to(torch_device)
```
Comment (vasqu) [inline tests/models/chameleon/test_modeling_chameleon.py:273]: Any reason that was changed?

File: tests/test_modeling_common.py [hunk 5]
```diff
@@ -4108,9 +4108,13 @@ def test_flash_attn_2_can_compile_with_attention_mask_None_without_graph_break(s
 
         assert not loss.isnan().any()
 
-    def flash_attention_padding_matches_padding_free_with_position_ids(
+    def attention_mask_padding_matches_padding_free_with_position_ids(
```
Comment (vasqu) [inline tests/test_modeling_common.py:4128]: Not a fan of the rename tbh, would be nice if we could keep flash attn still in the name.

File: tests/test_modeling_common.py [hunk 6]
```diff
@@ -4122,20 +4126,30 @@ def flash_attention_padding_matches_padding_free_with_position_ids(
         }
 
         for model_class in self.all_generative_model_classes:
-            if not getattr(model_class, support_flag[attn_implementation]):
+            if attn_implementation != "eager" and not getattr(model_class, support_flag[attn_implementation]):
```
Comment (vasqu) [inline tests/test_modeling_common.py:4146]: #39576 also addresses this :D

File: tests/test_modeling_common.py [hunk 7]
```diff
@@ -4122,20 +4126,30 @@ def flash_attention_padding_matches_padding_free_with_position_ids(
...
                 self.skipTest(f"{model_class.__name__} does not support {attn_implementation}")
 
+            # can't infer if new attn mask API is supported by assume that only model with attention backend support it
+            if not model_class._supports_attention_backend:
+                self.skipTest(f"{model_class.__name__} does not support new attention mask API")
```
Comment (vasqu) [inline tests/test_modeling_common.py:4151]: I have a feeling that will skip a lot of previously run tests especially those that are in the refactoring process. Could you cross-check?

File: tests/test_modeling_common.py [hunk 8]
```diff
@@ -4122,20 +4126,30 @@ def flash_attention_padding_matches_padding_free_with_position_ids(
...
+
+            if model_class._is_stateful:  # non-transformer models most probably have no packing support
+                self.skipTest(f"{model_class.__name__} doesn't support packing!")
+
             config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
+            if config.is_encoder_decoder:
+                self.skipTest("Model is an encoder-decoder")
```
Comment (vasqu) [inline tests/test_modeling_common.py:4158]: Might be doable in the future but yea not for now

File: tests/test_modeling_common.py [hunk 9]
```diff
@@ -4147,11 +4161,14 @@ def flash_attention_padding_matches_padding_free_with_position_ids(
             with tempfile.TemporaryDirectory() as tmpdirname:
                 model.save_pretrained(tmpdirname)
 
-                # ensure left padding, to adapt for some models
+                # Drop all keys exceot for the minimal set. Hard to manipulate with multimodals/head_mask/etc
```
Comment (vasqu) [inline tests/test_modeling_common.py]:
```suggestion
                # Drop all keys except for the minimal set. Hard to manipulate with multimodals/head_mask/etc
```

File: tests/test_modeling_common.py (hunk 1 above)
Comment (vasqu) [reply tests/test_modeling_common.py:4218]: We have a better assert right below it, so should be fine

File: tests/models/gemma3/test_modeling_gemma3.py [hunk 10]
```diff
@@ -143,6 +143,14 @@ def test_eager_matches_fa2_generate(self):
     def test_multi_gpu_data_parallel_forward(self):
         pass
 
+    @unittest.skip("Gemma3 applies key/query norm which doesn't work with packing")
+    def test_eager_padding_matches_padding_free_with_position_ids(self):
+        pass
```
Comment (vasqu) [inline tests/models/gemma3/test_modeling_gemma3.py:148]: Hmm, because the diffs get higher, don't get the immediate reason there :D

**Review by vasqu — COMMENTED** [review]
Just some comments on the attention side of things, a lot of nits so sorry about that ^^'

File: tests/models/paligemma/test_modeling_paligemma.py [hunk 11]
```diff
@@ -290,6 +290,14 @@ def test_feed_forward_chunking(self):
     def test_flash_attention_2_padding_matches_padding_free_with_position_ids(self):
         pass
 
+    @unittest.skip("Palogemma position ids are 1 indexed")
```
Comment (molbap) [inline tests/models/paligemma/test_modeling_paligemma.py]: nits :p (and we should find a way to fix this indexing, ugh)
```suggestion
    @unittest.skip("Paligemma position ids are 1 indexed")
```

File: tests/test_modeling_common.py (hunk 6 above)
Comment (molbap) [reply tests/test_modeling_common.py:4146]: indeed haha, no matter!

File: tests/models/paligemma/test_modeling_paligemma.py (hunk 11 above)
Comment (zucchini-nlp) [reply tests/models/paligemma/test_modeling_paligemma.py]: True

File: tests/test_modeling_common.py (hunk 5 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4128]: we can do separate tests for FA2 only and other attentions with packing

File: tests/models/chameleon/test_modeling_chameleon.py (hunk 4 above)
Comment (zucchini-nlp) [reply tests/models/chameleon/test_modeling_chameleon.py:273]: at first to ensure packing works with mutlimodal data. But then I decided to limit tests with text-only input 😄 

This change is here only for correctness now, because the mask is prepared based on actual input ids

File: tests/models/emu3/test_modeling_emu3.py (hunk 3 above)
Comment (zucchini-nlp) [reply tests/models/emu3/test_modeling_emu3.py]: oops, 1 is a typo. The order matters because a few lines above `pad_token` is re-assigned to image token

File: tests/test_modeling_common.py (hunk 5 above)
Comment (vasqu) [reply tests/test_modeling_common.py:4128]: Fair, kinda forgot that sdpa etc can also now use packing

Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4128]: I feel like it became too redundant after making them separate, I'd prefer to keep it as is and just add a small docstring about what we are testing

File: tests/test_modeling_common.py (hunk 7 above)
Comment (zucchini-nlp) [reply tests/test_modeling_common.py:4151]: It is currently skipping all the failing tests which either don't have new attention API or are super old like Xmod/Roberta etc (~25 models)

Though I think after your PR this will fail for 5 Bert-style generative models. Maybe we can skip them in model's test files until attention refactor PR

Comment (vasqu) [reply tests/test_modeling_common.py:4151]: Hmm, the issue is that the attention backend also needs the kwargs right? But since we have 3 different types of attentions, I need to make another PR there to properly do this. 

Let's keep it for now, I'll keep it in mind during my refactors tho

**Review by vasqu — APPROVED** [review]
LGTM, keeping the supports attn backend in mind for myself
